ci: Use dafny-reportgenerator instead of Python script for verification time limits

atomb opened the pull request:
*Description of changes:*

The `verify-macos.yaml` workflow previously used a custom Python script to identify whether any individual proof ran for more than a given amount of time. The `dafny-reportgenerator` tool can do this, and also some deeper verification stability analysis.

This PR updates the verification workflow to use `dafny-reportgenerator` instead, and bounds the cost of verification using more deterministic solver "resource count" metric instead of time.

It also adds two extra rounds of verification and assesses the stability of verification by calculating the coefficient of variation of the resource counts over the three total iterations. This increases the runtime of the job from about 7m to about 17m. We could investigate different forms of parallelism if this is a problematic increase.

The invocation of `dafny-reportgenerator` includes the `--allow-different-outcomes` flag, which skips the process of flagging proofs that succeed with some random seeds and fail with others, since a few of the proofs in this repo would otherwise fail. Ultimately, it probably makes most sense to refactor those proofs to be more stable, but the current set of flags at least puts in guard rails to prevent any reduction in proof stability. To start the refactoring process, this PR also modifies one function very slightly to improve the stability of one of the lemmas.

*Squash/merge commit message, if applicable:*

Same as description above.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

atomb committed:
Use dafny-reportgenerator instead of Python script

atomb committed:
Install .NET 6.0

atomb committed:
Go back to Dafny 3.5.0

atomb committed:
Change where to look for results

atomb committed:
Do stability analysis

atomb committed:
Revert "Do stability analysis"

This reverts commit 4f2972ff51caed3b8ec1fb3d384d58508a4e614f.

Need a newer Dafny to do this

atomb committed:
Run Dafny directly

This will be needed for stability analysis, as the MSBuild approach
doesn't allow multiple custom command-line options.

atomb committed:
Revert to `dotnet build` for verification

atomb committed:
Look for CSVs in a different place

atomb committed:
Minimize differences

atomb committed:
Use resource count limit instead of time limit

atomb committed:
Revert "Use resource count limit instead of time limit"

This reverts commit 72d0397b74b7ddd0d17b11ac666d827ff19283c3.

Checking resource limits won't work until the CI uses Dafny 3.6 or
newer.

atomb committed:
Add comment and bump timeout ever so slightly

seebees commented on .github/workflows/verify-macos.yml:
```diff
@@ -14,14 +14,24 @@ jobs:
       - name: Verify Dafny code
         # Currently, test depends on src, so verifying test will also verify src
-        run: dotnet build -t:VerifyDafny -p:VerifyDafnyJobs=2 -p:TestVerifyOverride="verificationLogger:trx" test
+        run: dotnet build -t:VerifyDafny -p:VerifyDafnyJobs=2 -p:TestVerifyOverride="verificationLogger:csv" test
```
Why change it to `csv`? I think that this file is uploaded as an artifact to CodeBuild

atomb replied:
CSV is what `dafny-reportgenerator` needs as input. It's possible to enable both formats of output, but I think we'd have to invoke `dafny` directly instead of using `dotnet build` for that to work, since I believe it's only possible to specify one `-p:TestVerifyOverride` argument.

On a related note, I've tried verifying the code base with Dafny 3.7.1, on macOS, and it succeeds for me when I directly invoke Dafny but not when I use `dotnet build`, strangely enough. With `dotnet build`, it eventually hangs, using no CPU. I think I've seen some concurrency issues in `dotnet` on macOS in other contexts to, and wonder whether something similar is showing up here.

The bottom line is that it might be worth running `dafny` directly here. That could solve both issues.

seebees commented on .github/workflows/verify-macos.yml:
```diff
@@ -14,14 +14,24 @@ jobs:
-        run: MAX_VERIFICATION_DURATION_SECONDS=40 python3 verification-times-from-trx.py test/TestResults/*.trx
+      - name: Check durations
+        if: always()
+        run: dafny-reportgenerator summarize-csv-results --max-duration-seconds 45 test/TestResults/*.csv
```
Why 40 -> 45?

atomb replied:
On at least one of the runs I did, one of the proofs took a tiny bit over 40s. I suspect that, when we take the next step to enable stability analysis, we'll find that one to be unstable and probably in need of refactoring to improve stability. For the time being, though, it seemed worth bumping the timeout slightly to get things to go through.

atomb committed:
Continue to generate `.trx` files

This requires moving away from `dotnet build`, unfortunately.

atomb committed:
Fix path

atomb committed:
Bound on resource count instead of time

atomb committed:
Bump solver resource count limit

atomb committed:
Bump to Dafny 3.7.1

atomb committed:
Tweak comment and job name

atomb committed:
Use `dotnet tool install` for everything

atomb committed:
Change only one variable at a time

atomb committed:
Revert "Change only one variable at a time"

This reverts commit 1d510d34cd640a6464eceeac961c3ae3908b215d.

atomb committed:
Revert "Use `dotnet tool install` for everything"

This reverts commit a11e7d1afd519ab69c346ba55298ed038927f12f.

We would need to install Z3 separately for this to work.

atomb committed:
Use dafny-reportgenerator v1.2.0

atomb committed:
Try /randomSeedIterations

atomb committed:
Use more cores; check stability results

GitHub's macOS runners have 3 cores, so 4 threads is probably
reasonable.

Also adds a /timeLimit on the stability analysis portion, because
different random seeds can sometimes lead to dramatically different
proof durations.

atomb committed:
Make one proof more stable

atomb committed:
Update comments; test two limits separately

atomb committed:
Merge branch 'mainline' into dafny-reportgenerator
